### PR TITLE
feature: status returns the function name when called with -u parameter

### PR DESCRIPTION
**Pull request by Markcial**
## Description

Added a `--current-function`, `-u` parameter for the `status` function, if the function  exists on the parser the function name is sent to the out stream, when missing a "Not a function" message is sended instead.

It's unclear what to do when the function is not found.
Maybe the terminal should: 
  - display message and exit code 0
  - no message and exit code 1
  - display message and exit code 1
  
Fixes issue #1743

## TODOs:

- [ ] Changes to fish usage are reflected in user documentation/manpages.
- [ ] Tests have been added for regressions fixed
- [ ] User-visible changes noted in CHANGELOG.md

### CODE REVIEW

File: src/builtin.cpp [hunk 1]
```diff
@@ -2451,7 +2453,7 @@ static int builtin_status(parser_t &parser, io_streams_t &streams, wchar_t **arg
     /// the non-flag subcommand form. While these flags are deprecated they must be supported at
     /// least until fish 3.0 and possibly longer to avoid breaking everyones config.fish and other
     /// scripts.
-    const wchar_t *short_options = L":cbilfnhj:t";
+    const wchar_t *short_options = L":bcfhij:lnut";
```
Comment (krader1961) [inline src/builtin.cpp]: The short and long flag forms are deprecated; see the second to last sentence in the man page. We do not want to introduce new ones.

File: src/parser.cpp [hunk 2]
```diff
@@ -440,6 +440,10 @@ const wchar_t *parser_t::is_function() const {
     return result;
 }
 
+const wchar_t *parser_t::get_function() {
```
Comment (krader1961) [inline src/parser.cpp]: I don't like needlessly long symbol names but `get_function_name()` or `get_func_name()` better describes the purpose of the function.

**Comment by krader1961** [comment]
As for what to do if not in function context I prefer your second option. However, `status current-filename`, and possibly others, has already established the pattern you implemented so we should probably continue in that vein. But I'd like to see what others think before this is merged as now is a good time to change the pattern. Note that if we add a `--level -1` special-case as described below to report all levels a calling context which was not a function would simply report an empty string rather than `Not a function`.

What I would like to see happen next (via a new issue) is for a flag to be introduced that specifies which level in the call stack to report on. With zero, or not specifying the flag, being equivalent to the "current" level. And `1` being the calling context, etc. With the special value `-1` meaning to report all levels, one per line. This would then make it possible do more interesting things like selectively report the context from which the current code was called without being limited to what `status print-stack-trace` provides. With that ability `status print-stack-trace` could be reimplemented via a function wrapper around the builtin. Not that we would do so but the basic idea opens up other potentially useful error reporting and debugging mechanisms.

File: src/builtin.cpp (hunk 1 above)
Comment (Markcial) [inline src/builtin.cpp]: How I should add the `current-function` parameter without messing up the deprecated style? Is there any sample on how to base my rewrite? With the symbol name is enough I will search it and look how it is done and apply on the PR changes.

Comment (krader1961) [inline src/builtin.cpp]: Simply add the new subcommand to the `status_enum_map` like you've already done and omit defining flags as an alternative. In general we're trying to switch `history`, `status`, etc. from a `command --subcommand` idiom to `command subcommand` (like `git log`).

Legacy subcommands need to continue to allow the flag form for backward compatibility but we don't want to define flags for new subcommands.

**Comment by Markcial** [comment]
I can refactor all the other blocks to report error codes when the return is missing or whatever specified for the next release, and i can do that on this same PR if you want.

The former part you are talking about is not much trouble, but the later would require a new issue, it's true, mainly because i might be not able to help to that level of depth (yet).

So if more people want to specify or decide if we switch the behaviour on the other status methods i would be able to apply those changes on this same PR then.

**Comment by krader1961** [comment]
Since no one else has offered an opinion I went ahead and squash merged this as commit 980af4aa5. Thanks.

**Pull request closed without merging**